File: aungnanda_oo/q6_MissingInteger.py
# Question 6: MissingInteger

# Given an integer n and a sorted array of integers of size n - 1 which contains all but one of the integers in the range 1 - n, find the missing integer.

# Examples:

# Input Array: [1, 2, 3, 4, 6, 7]
# Input n: 7
# Output: 5

# Input Array: [1]
# Input n: 2
# Output: 2

# Input Array: [1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12]
# Input n: 12
# Output: 9

def MissingInteger(arr, n):
    # Sum of 1 to n = n * (n + 1) / 2
    expected_sum = n * (n + 1) // 2
    actual_sum = sum(arr)
    return expected_sum - actual_sum

# Time Complexity: O(n)
# Space Complexity: O(1)

test_cases = [
    ([1, 2, 3, 4, 6, 7], 7),
    ([1], 2),
    ([1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12], 12)
]
for arr, n in test_cases:
    print(MissingInteger(arr, n))


# Spent a total of 40 mins on this question, its just math thing :)

File: homework1/tom_nguyen/q6_MissingInteger.cpp
#include <iostream>
#include <vector>
using namespace std;

//uses the sum formula to find the missing integer in O(n)
int missingInteger(const vector<int>& arr, int n) {
    int expectedSum = n * (n + 1) / 2;
    int actualSum = 0;
    for (int num : arr) {
        actualSum += num;
    }
    return expectedSum - actualSum;
}

int main() {
    vector<int> t1 = {1, 2, 3, 4, 6, 7};
    cout << missingInteger(t1, 7) << endl; // 5

    vector<int> t2 = {1};
    cout << missingInteger(t2, 2) << endl; // 2

    vector<int> t3 = {1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 12};
    cout << missingInteger(t3, 12) << endl; // 9

    return 0;
}

//time spent: 5 mins
